File: typescript/packages/core/src/thread/fork.ts
import type { BranchId, EventId } from "../log";
import { knownEvents } from "../reduce";
import { err, ok, type Result } from "../result";
import { isRefusal, ownerContext } from "../sandbox/context";
import {
  lookupSandbox,
  release,
  resolvePending,
  sameProvider,
  settle,
} from "../sandbox/ledger";
import type {
  RestoreFailure,
  Sandbox,
  SandboxSession,
  SnapshotData,
} from "../sandbox/protocol";
import type { LogStore, ResourceRow, ResourceState, Writer } from "../store";
import { type LogError, logError } from "../verify/error";

// fork(): a new branch restored into an isolated sandbox, and only then
// visible. The child's ledger row and its operation key are durable before the restore call; a
// lost response resolves by that key, never by creating again.

export type KnowledgePolicy = "pinned" | "current";

export type ForkInput = {
  readonly parent: BranchId;
  readonly point: EventId;
  readonly child: BranchId;
  readonly knowledge: KnowledgePolicy;
  readonly holderId: string;
};

/** The snapshot event `point` names on the parent's chain, or no_snapshot_boundary. */
async function snapshotAt(log: LogStore, input: ForkInput) {
  const parent = await log.read(input.parent);
  if (!parent.ok) return parent;
  const event = knownEvents(parent.value).find(
    (e) => e.event_id === input.point,
  );
  if (event === undefined)
    return err(
      logError(
        "no_snapshot_boundary",
        `no event ${input.point} on this branch`,
      ),
    );
  return ok({
    seq: event.seq,
    snapshot: event.type === "snapshot" ? event.data : undefined,
  });
}

/** Creates the child branch, restores its sandbox, and makes it ready; else fails it cleanly. */
export async function forkBranch(
  log: LogStore,
  sandbox: Sandbox | undefined,
  input: ForkInput,
): Promise<Result<void, LogError>> {
  const at = await snapshotAt(log, input);
  if (!at.ok) return at;
  const writer = await log.beginFork({
    parent: input.parent,
    atSeq: at.value.seq,
    branch: input.child,
    holderId: input.holderId,
  });
  if (!writer.ok) return writer;
  const snapshot = at.value.snapshot;
  if (snapshot === undefined)
    throw new Error("an eligible point is a snapshot");
  const fail = async (code: LogError["code"], message: string) => {
    const failed = await log.failFork(writer.value);
    return failed.ok ? err(logError(code, message, at.value.seq)) : failed;
  };
  if (sandbox?.info.provider !== snapshot.provider)
    return fail(
      "snapshot_restore_failed",
      `no ${snapshot.provider} sandbox adapter to restore with`,
    );
  const restored = await restore(log, writer.value, sandbox, snapshot);
  if (!restored.ok)
    return restored.error.code === "stale_epoch" ||
      restored.error.code === "writer_poisoned"
      ? restored
      : fail(restored.error.code, restored.error.message);
  // If this fails the lease was lost before the child became visible; the fork's next owner
  // reclaims it (LogStore.reclaimFork) and finds the live row in the ledger.
  return log.finishFork(writer.value, {
    sandboxId: restored.value.id,
    knowledgePolicy: input.knowledge,
  });
}

/**
 * The restore under a ledger row. A lost response (`unavailable`) resolves by lookup: a found
 * sandbox is used, a final not_found fails the fork, anything else leaves the row unknown for an
 * operator (`resource_unknown`). A typed failure releases whatever a lookup still finds.
 */
async function restore(
  log: LogStore,
  writer: Writer,
  sandbox: Sandbox,
  snapshot: SnapshotData,
): Promise<Result<SandboxSession, LogError>> {
  const ledger = log.ledger;
  const row = await ledger.begin(writer, "sandbox", sandbox.info.provider);
  if (!row.ok) return row;
  // The adapter fences this context at its real dispatch point, after any queueing: a
  // creator that lost the lease meanwhile creates nothing.
  const made = await sandbox.restore(
    snapshot.snapshot_id,
    snapshot.manifest_hash,
    row.value.operation_key,
    ownerContext(writer),
  );
  if (made.ok) {
    const live = await ledger.live(
      writer,
      row.value.resource_id,
      made.value.id,
      null,
    );
    return live.ok ? ok(made.value) : live;
  }
  if (isRefusal(made.error))
    return err(logError("stale_epoch", made.error.message));
  return recover(log, writer, sandbox, row.value, made.error);
}

async function recover(
  log: LogStore,
  writer: Writer,
  sandbox: Sandbox,
  row: ResourceRow,
  failure: RestoreFailure,
): Promise<Result<SandboxSession, LogError>> {
  const ledger = log.ledger;
  const answer = await lookupSandbox(
    sandbox,
    row.operation_key,
    ownerContext(writer),
  );
  if (!answer.ok) return answer;
  const settled = await settle(
    ledger,
    writer,
    row,
    answer.value,
    sandbox.info.lookup.create === "final",
    (s) => s.id,
  );
  if (!settled.ok) return settled;
  const { state } = settled.value;
  if (state === "unknown")
    return err(
      logError(
        "resource_unknown",
        `the child sandbox of operation ${row.operation_key} can't be established`,
      ),
    );
  if (settled.value.state === "live" && failure.code === "unavailable")
    return ok(settled.value.value);
  if (settled.value.state === "live") {
    const released = await release(
      ledger,
      writer,
      sandbox,
      row,
      settled.value.value,
    );
    if (!released.ok) return released;
  }
  const code =
    failure.code === "unavailable" ? "snapshot_restore_failed" : failure.code;
  return err(logError(code, failure.message));
}

/**
 * After a crash mid-fork the fork's creator owns cleanup: it retakes the
 * forking branch, settles each ledger row the fork recorded (a pending one by lookup, never by
 * creating again; a live or failed one by release), and marks the branch fork_failed.
 */
export async function recoverFork(
  log: LogStore,
  sandbox: Sandbox,
  branch: BranchId,
  holderId: string,
): Promise<Result<readonly ResourceState[], LogError>> {
  const rows = await log.ledger.rows(branch);
  if (!rows.ok) return rows;
  // Only the rows' own provider can prove anything about them; refuse before touching them.
  for (const row of rows.value) {
    const same = sameProvider(sandbox, row);
    if (!same.ok) return same;
  }
  const writer = await log.reclaimFork(branch, holderId);
  if (!writer.ok) return writer;
  const states: ResourceState[] = [];
  for (const row of rows.value) {
    const settled =
      row.state === "pending"
        ? await resolvePending(log.ledger, writer.value, sandbox, row)
        : row.state === "live" || row.state === "release_failed"
          ? await release(log.ledger, writer.value, sandbox, row)
          : ok(row.state);
    if (!settled.ok) return settled;
    states.push(settled.value);
  }
  const failed = await log.failFork(writer.value);
  return failed.ok ? ok(states) : failed;
}

/**
 * Startup recovery for forks: every branch a crash left `forking` whose lease has lapsed is
 * failed and its ledger rows settled. A fork is never resumed. Returns the branches it failed.
 */
export async function recoverForks(
  log: LogStore,
  sandbox: Sandbox,
  holderId: string,
): Promise<Result<readonly BranchId[], LogError>> {
  const forking = await log.forkingBranches();
  if (!forking.ok) return forking;
  const failed: BranchId[] = [];
  for (const branch of forking.value) {
    const done = await recoverFork(log, sandbox, branch, holderId);
    if (done.ok) failed.push(branch);
    else if (done.error.code !== "branch_busy") return done;
  }
  return ok(failed);
}

File: python/src/threads/thread/fork.py
"""The fork operation: a new branch of the same thread at an eligible
snapshot, restored into an isolated sandbox that the child owns.

Only a quiescent, unexpired snapshot event is a fork point. The child is `forking` (unlisted)
until its `fork` event is written; the restore's ledger row is written before the provider
call. On any failure the child becomes `fork_failed` and what the fork created is released,
except a restore whose outcome the adapter can't establish: that row stays `unknown` for an
operator (resource_unknown). The parent is never written.
"""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

from threads.log import (
    BranchId,
    Event,
    EventId,
    ForkEvent,
    ParseError,
    SnapshotEvent,
)
from threads.reduce import Fold
from threads.result import Err, Ok
from threads.sandbox.ledger import Fenced, Tracked, abandon, acquire, release_session
from threads.sandbox.protocol import Sandbox, SandboxError, session_lookup
from threads.store import SqliteStore, Writer
from threads.store.worker import Clock

if TYPE_CHECKING:
    from pydantic import JsonValue

type KnowledgePolicy = Literal["pinned", "current"]


@dataclass(frozen=True, slots=True)
class ForkAt:
    parent: BranchId
    """The branch whose resolved chain holds the snapshot."""
    point: EventId
    """The snapshot event."""
    child: BranchId
    knowledge: KnowledgePolicy = "pinned"


async def fork_branch(
    store: SqliteStore, sandbox: Sandbox | None, at: ForkAt, holder_id: str, clock: Clock
) -> Ok[Writer] | Err[ParseError]:
    """Forks at `at.point` and returns the child's writer, holding its first lease."""
    eligible = await _restorable(store, sandbox, at, clock)
    if isinstance(eligible, Err):
        return eligible
    snap, sandbox = eligible.value
    begun = await store.begin_fork(snap.branch_id, snap.seq, at.child, holder_id, clock)
    if isinstance(begun, Err):
        return begun
    forking = begun.value
    context = store.context(forking.owner, clock)
    restore = Tracked(
        "sandbox",
        lambda key: sandbox.restore(snap.data.snapshot_id, snap.data.manifest_hash, key, context),
        *session_lookup(sandbox, context),
        lambda session: session.id,
    )
    restored = await acquire(store.ledger, forking.owner, snap.data.provider, restore, clock)
    if isinstance(restored, Err):
        await store.fail_fork(forking.row.branch_id)
        return Err(_failure(restored.error, snap.seq))
    row, session = restored.value
    data: dict[str, JsonValue] = {
        "reason": "snapshot",
        "sandbox_id": session.id,
        "knowledge_policy": at.knowledge,
    }
    finished = await store.finish_fork(forking, data, clock)
    if isinstance(finished, Ok) and finished.value is not None:
        return Ok(finished.value)
    await release_session(Fenced(store.ledger, forking.owner, context, clock), row, session)
    await store.fail_fork(forking.row.branch_id)
    return finished if isinstance(finished, Err) else Err(_not_runnable(snap.seq))


async def recover_forks(
    store: SqliteStore, sandbox: Sandbox, holder_id: str, clock: Clock
) -> tuple[BranchId, ...]:
    """A fork is never resumed (spec/schema/README.md): each fork a crash left `forking` gives
    up every ledger row it wrote (released, or unknown for an operator) and becomes
    `fork_failed`. Returns the children it failed."""
    failed: list[BranchId] = []
    for owner in await store.interrupted_forks(holder_id, clock):
        by = Fenced(store.ledger, owner, store.context(owner, clock), clock)
        for row in await store.ledger.rows():
            if row.owner_branch_id == owner.branch_id:
                await abandon(by, sandbox, row)
        await store.fail_fork(owner.branch_id)
        failed.append(owner.branch_id)
    return tuple(failed)


async def _restorable(
    store: SqliteStore, sandbox: Sandbox | None, at: ForkAt, clock: Clock
) -> Ok[tuple[SnapshotEvent, Sandbox]] | Err[ParseError]:
    read = await store.read(at.parent, clock())
    if isinstance(read, Err):
        return read
    eligible = fork_point(read.value.fold, at.point)
    if isinstance(eligible, Err):
        return eligible
    snap = eligible.value
    if sandbox is None or sandbox.info.provider != snap.data.provider:
        message = f"no sandbox adapter for provider {snap.data.provider}"
        return Err(ParseError("sandbox_required", message, snap.seq))
    return Ok((snap, sandbox))


def fork_point(fold: Fold, point: EventId) -> Ok[SnapshotEvent] | Err[ParseError]:
    """The snapshot event `point` if it is an eligible fork point on this chain now."""
    event = next((e for e in fold.events if e.event_id == point), None)
    if not isinstance(event, SnapshotEvent):
        seq = None if event is None else event.seq
        return Err(ParseError("no_snapshot_boundary", f"{point} is not a snapshot event", seq))
    expires = event.data.expires_at
    if expires is not None and expires <= fold.now:
        return Err(ParseError("snapshot_expired", f"snapshot {point} has expired", event.seq))
    if (event.seq, event.event_id) not in fold.fork_points:
        message = f"snapshot {point} was not taken at a quiescent boundary"
        return Err(ParseError("no_snapshot_boundary", message, event.seq))
    return Ok(event)


def knowledge_revision(events: tuple[Event, ...] | list[Event]) -> int | None:
    """The knowledge revision a branch searches as of: its fork snapshot's
    under `pinned`, None (the live corpus) under `current`, for a root, or for a snapshot
    taken without a knowledge binding."""
    forks = [e for e in events if isinstance(e, ForkEvent)]
    if not forks or forks[-1].data.knowledge_policy != "pinned":
        return None
    at_seq = forks[-1].seq - 1
    snap = next((e for e in events if e.seq == at_seq and isinstance(e, SnapshotEvent)), None)
    revision = None if snap is None else snap.data.knowledge_revision
    return revision if isinstance(revision, int) else None


def _failure(error: SandboxError | ParseError, seq: int) -> ParseError:
    if isinstance(error, ParseError):
        return error
    match error.code:
        case (
            "snapshot_expired" | "snapshot_missing" | "snapshot_manifest_mismatch"
        ) | "resource_unknown" as code:
            return ParseError(code, error.message, seq)
        case _:
            return ParseError("snapshot_restore_failed", error.message, seq)


def _not_runnable(seq: int) -> ParseError:
    # Only a repair child is not runnable, and a snapshot fork never is one.
    return ParseError("invalid_transition", "the forked child is not runnable", seq)
